### ISSUE TITLE
orderBy - crazy Array.prototype.sort behaviour in Chromium browsers - Cannot clear sort

### ISSUE THREAD

**Issue by DumboJetEngine** [angular/angular.js#11642]
I have found a bug that has been properly described here:

http://stackoverflow.com/questions/29716165/angularjs-orderby-crazy-array-prototype-sort-behaviour-in-chromium-browsers

This might be more of a Chromium bug, but you are all a big Google family, so I post it here. :)

Please let me know of any decisions/ideas/etc.
This bug made my life miserable.

----
**UPDATE:**
If this is not considered to be a bug (which would be too hard for me to swallow), then I would like a better way than this:
http://stackoverflow.com/a/29719391/964053
...and a better way than creating a custom filter in order to have a way to preserve the original order of the objects in an array.

That last one might not be too much for me to do, but I can't wrap my head around the current behavior of the `orderBy` filter on arrays of objects when `predicate===''`...
I have tried fixing it like so:

```
orderByFilter.$inject = ['$parse'];
function orderByFilter($parse) {
  return function(array, sortPredicate, reverseOrder) {
    if ((sortPredicate === undefined || sortPredicate === '') && isArray(array)) return array;
    if (!(isArrayLike(array))) return array;
```
...but on second thought, a custom filter would be a more maintainable for me solution.

**Comment by lgalfaso** [comment]
The simplest way to make the sort stable would be to add a schwartz transform, add the original position and add a tie breaker condition. The issue with this is that it would generate a lot of garbage that would put stress on the JS VM... With some fear to make `orderBy` implementation too big, an option would be to add a forth parameter to `orderBy` to ask it to use a stable sort using the alternative algorithm stated before

**Comment by gkalpak** [comment]
Wouldn't it make sense to just skip sorting (i.e. return the original array) if `sortPredicate` is `null`/`undefined` ? Currently, calling `orderBy(arr, null/undefined)` boils down to `arrCopy.sort(function () { return 0; })`, so I don't think we are breaking any expectation (unless someone was relying on the fact that calling `sort(function () { return 0; })` might sort items in arbitrary order).

Making `orderBy` consistently return the original array (or a copy of it) when the predicate is `null`/`undefined` is a good thing imo.
(It's not even a breaking change and it doesn't bloat `orderBy`.)

**Comment by lgalfaso** [comment]
It would also imply that orderBy(arr, null/undefined) and orderBy(arr,
'somePropertyUndefinedToAllElements') would behave different
>
>

**Comment by gkalpak** [comment]
tl;dr: I don't think this is a problem; do you ?

--
According to my understanding, `orderBy(arr, null/undefined)` means "I want you to sort by nothing; i.e. not sort at all", while `orderBy(arr, 'someProp')` means "I want you to sort by `someProp` (not caring what it's value in each item is and accepting the fact that is might be `undefined` or `42` or _xyz_ in **all** of them and I rely on the browser to do the sorting for me and I accept that browsers might have inconsistent implementations, which Angular can do nothing about)".
Too verbose, but you get the point (hopefully).

Basically, I suggest we differente the following cases: _sort by nothing_ and _sort by some property that has the same value in all items_
Currently, we are treating the former case as if it were the latter (and quite arbitrarily so, imo), but I don't think anyone was relying on the fact that sorting by `null`/`undefined` will produce the same results as sorting by a property that has the same value in all items (even if these results are inconsistent between browsers).

**Comment by DumboJetEngine** [comment]
If anyone cares for my thoughts, the way I see it, this will have many benefits:

- Reliability: I expect my libraries to be functioning the same way in all browsers. No developer likes surprises. I don't know why the JS specs are so soft on these issues, but I would like Angular (and every JS library I use) to work the same everywhere when possible (unless there is a serious reason not to, like performance issues). Keeping the array intact when predicate is `null/undefined/''` does help in that direction. Now if it comes to predicates that do not map to an existing property, I cannot imagine who would ever do that and for what reason.
- Performance: When `predicate===''` you can skip all the rest of the processing that Angular does (twice on each digest if I am not mistaken). And, statistically speaking, grids in an angular application usually are unsorted initially until the user asks for sorting. I don't have any benchmarks, but this could make a performance difference in many applications.
- Flexibility: In order to keep the original order (which seems to me like a reasonable and frequent requirement -- I wonder why no-one else hasn't discovered this before...) you don't have to do any of these: http://stackoverflow.com/a/29722904/964053, http://stackoverflow.com/a/29719391/964053.
- Also, I am not sure if you are aware of this, but this version of Angular: http://ajax.googleapis.com/ajax/libs/angularjs/1.2.8/angular.js (and many others before it) works like that because it has this line:

        if (!sortPredicate) return array;
So, this is not new behavior either.
I am not sure why this was removed completely and was not improved instead.

Another option would be to add another keyword predicate like you do with '+' that will denote the `DO NOT SORT`. E.g. to skip sort when `predicate==='.'`.

**Comment by lgalfaso** [comment]
Now that ed3a33a063f09d7ca356d15c278d95ad82e680a0 landed, this should no longer be an issue
